source/extensions/watchdog/abort_action/abort_action.cc

```diff
@@ -0,0 +1,69 @@
+#include "extensions/watchdog/abort_action/abort_action.h"
+
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+
+#ifndef WIN32
+#include <sys/types.h>
+#include <csignal>
+#endif
+
+#include "common/common/logger.h"
+#include "common/protobuf/utility.h"
+#include "envoy/thread/thread.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Watchdog {
+namespace AbortAction {
+namespace {
+#ifdef __linux__
+pid_t toPlatformTid(int64_t tid) { return static_cast<pid_t>(tid); }
+#elif defined(__APPLE__)
+uint64_t toPlatformTid(int64_t tid) { return static_cast<uint64_t>(tid); }
+#endif
+} // namespace
+
+AbortAction::AbortAction(
+    envoy::extensions::watchdog::abort_action::v3alpha::AbortActionConfig& config,
+    Server::Configuration::GuardDogActionFactoryContext& /*context*/)
+    : config_(config){};
+
+void AbortAction::run(
+    envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent /*event*/,
+    const std::vector<std::pair<Thread::ThreadId, MonotonicTime>>& thread_last_checkin_pairs,
+    MonotonicTime /*now*/) {
+
+  if (thread_last_checkin_pairs.empty()) {
+    ENVOY_LOG_MISC(warn, "Watchdog AbortAction called without any thread.");
+    return;
+  }
+
+  int64_t raw_tid = thread_last_checkin_pairs[0].first.getId();
+
+#ifdef WIN32
+  // TODO(kbaichoo): add support for this with windows.
+  ENVOY_LOG_MISC(error, "Watchdog AbortAction is unimplemented for Windows.");
+#else
+  // Assume POSIX-compatible system and signal to the thread.
+  ENVOY_LOG_MISC(error, "Watchdog AbortAction sending abort signal to thread with tid {}.",
+                 raw_tid);
+
+  if (kill(toPlatformTid(raw_tid), SIGABRT) == 0) {
+    // Successfully sent signal, sleep for wait_duration.
+    absl::SleepFor(absl::Milliseconds(PROTOBUF_GET_MS_OR_DEFAULT(config_, wait_duration, 0)));
+  } else {
+    // Failed to send the signal, abort?
+    ENVOY_LOG_MISC(error, "Failed to send signal to tid {}", raw_tid);
+  }
+
+  // Abort from the action since the signaled thread hasn't yet crashed the process.
+  PANIC(fmt::format("Failed to kill thread with id {}, aborting from Watchdog AbortAction instead.",
+                    raw_tid));
```

Comment:
I haven't kept fully up to date on this, but isn't the default action at the end of the action chain to terminate the process? Do we need this or would the action chain just end up aborting anyway?

Reply:
Yes, you're right. I discussed this with antonio up above. I've copied the reasoning below (copied from https://github.com/envoyproxy/envoy/pull/12860#discussion_r480353221)

There are 3 reasons I think this [having a panic in the action] would be handy (vs deferring to the underlying watchdog default actions[its panic there]):

- Easier to see where the failure occurred, and that signaling failed without having to jump around as much to understand the behavior.
- It's also keeps this self-contained if there were changes on the underlying system behavior in watchdog kill / multikill default
- Allows flexibility for using this in places where there isn't a default PANIC afterwards

Given that we're planning to make this also a default action on platform where it's supported I'd likely make the following change in the next PR:
- If we're doing kill or multikill, then install the abort action as the final watchdog action for either of those events.

I could possibly see when this has full support across platforms that we'd remove the watchdog's default panic, since it'd end up being dead code.

Reply:
OK that make sense. Can you summarize this comment in the code? Thank you.

/wait

Reply:
Done.